datasetComparison/src/main/scala/za/co/absa/hermes/datasetComparison/DatasetComparison.scala

```diff
@@ -0,0 +1,282 @@
+      passedCount,
+      selector,
+      resultDF,
+      diffCount,
+      cliOptions.rawOptions
+    )
+  }
+
+  /**
+   * Creates DataFrame that has the original data and differences that were found in error column.
+   *
+   * @param path Path where the difference will be written to
+   * @param dataFrames Pair of relative complements of reference and actual data
+   */
+  private def createDiffDataFrame(path: String,
+                                  dataFrames: ComparisonPair[DataFrame]): DataFrame = {
+    val joinedData: DataFrame = joinTwoDataFrames(dataFrames)
+
+    // Flatten data
+    val flatteningFormula = HelperFunctions.flattenSchema(dataFrames.reference)
+    val flatExpected: DataFrame = dataFrames.reference.select(flatteningFormula: _*)
+    val flatActual: DataFrame = dataFrames.actual.select(flatteningFormula: _*)
+
+    val joinedFlatDataWithoutErrCol: DataFrame = joinTwoDataFrames(ComparisonPair(flatExpected, flatActual))
+    val joinedFlatDataWithErrCol = joinedFlatDataWithoutErrCol.withColumn(config.errorColumnName, lit(Array[String]()))
+
+    val columns: Array[String] = flatExpected.columns.filterNot(_ == config.comparisonUniqueId)
+    val flatDataWithErrors: DataFrame = findDifferences(columns, joinedFlatDataWithErrCol)
+
+    // Using the hash key, join the original data and error column from the flat data.
+    joinedData.as("df1")
+      .join(flatDataWithErrors.as("df2"), Seq(config.comparisonUniqueId))
+      .select("df1.*", s"df2.${config.errorColumnName}")
+      .drop(config.comparisonUniqueId)
+  }
+
+  /**
+   * Performs a check if the schemas of two data frames are actually the same.
+   *
+   * @param testedDf Comparison pair of two DataFrames whose schema will be tested
+   */
+  private def checkSchemas(testedDf: ComparisonPair[DataFrame]): Unit = {
+    val expectedSchema: StructType = getSchemaWithoutMetadata(testedDf.reference.schema)
+    val actualSchema: StructType = getSchemaWithoutMetadata(testedDf.actual.schema)
+
+    if (!SchemaUtils.isSameSchema(expectedSchema, actualSchema)) {
+      val diffSchema = SchemaUtils.diffSchema(expectedSchema, actualSchema) ++
+        SchemaUtils.diffSchema(actualSchema, expectedSchema)
+      throw SchemasDifferException(cliOptions.referenceOptions.path, cliOptions.newOptions.path, diffSchema.mkString("\n"))
+    }
+  }
+
+  /**
+   * Handles duplicates in a sense that this method looks for them. Then based on the configuration if found, throws an
+   * error, does deduplication or just passes through.
+   *
+   * @param dfsWithKey DataFrame pair where both have appended unique key
+   * @return
+   */
+  private def handleDuplicates(dfsWithKey: ComparisonPair[DataFrame]): (ComparisonPair[Long], ComparisonPair[DataFrame]) = {
+    def write(df: DataFrame, duplicates: DataFrame, path: String): Unit = {
+      df.alias("original")
+        .join(duplicates, Seq(config.comparisonUniqueId), "inner")
+        .select("original.*")
```

Comment:
Does this contain the special `comparisonUniqueId` column?

Reply:
Still does, doesn't. I will fix it